dataprep/clean/clean_address.py

```diff
@@ -0,0 +1,473 @@
+            address to be successfully cleaned.
+                - 'house_number': ('1234')
+                - 'street_prefix': ('N', 'North')
+                - 'street_name': ('Main')
+                - 'street_suffix': ('St', 'Avenue')
+                - 'apartment': ('Apt 1')
+                - 'building': ('Staples Center')
+                - 'city': ('Los Angeles')
+                - 'state': ('CA', 'California')
+                - 'zipcode': ('57903')
+
+            (default: ("house_number", "street_name"))
+
+     Examples
+     --------
+
+     >>> df = pd.DataFrame({"address":["123 pine avenue", "NULL"]})
+     >>> validate_address(df["address"])
+     0    True
+     1    False
+     Name: address, dtype: bool
+    """
+
+    if isinstance(x, pd.Series):
+        return x.apply(_check_address, args=(must_contain, False))
+
+    return _check_address(x, must_contain, False)
+
+
+def _format_address(
+    address: str, output_format: str, must_contain: Tuple[str, ...], split: bool, errors: str
+) -> Any:
+    """
+    Function to transform an address instance into the desired format
+
+    The last component of the returned tuple contains a code indicating how the
+    input value was changed:
+        0 := the value is null
+        1 := the value could not be parsed
+        2 := the value is cleaned and the cleaned value is DIFFERENT than the input value
+        3 := the value is cleaned and is THE SAME as the input value (no transformation)
+    """
+    address_dict, status = _check_address(address, must_contain, True)
+    outputs = _address_dict_to_string(address_dict, output_format, split)
+
+    if status == "null":
+        return (np.nan,) * len(_get_column_names(output_format, split)) + (0,)
+
+    elif status == "unknown":
+        if errors == "raise":
+            raise ValueError(f"unable to parse value {address}")
+        return tuple(
+            np.nan if not value else value if errors == "ignore" else np.nan for value in outputs
+        ) + (1,)
+
+    if len(outputs) == 1 and address == outputs[0]:
+        code = 3
+    else:
+        code = 2
+    return tuple(np.nan if not value else value for value in outputs) + (code,)
+
+
+def _check_address(address_str: str, must_contain: Tuple[str, ...], clean: bool) -> Any:
+    """
+    Finds the index of the given country in the DATA dataframe.
+
+    Parameters
+    ----------
+    address_str
+        string containing the address value being cleaned
+    must_contain
+        A tuple containing parts of the address that must be included for the
+         address to be successfully cleaned
+    clean
+        If True, a tuple (index, status) is returned.
+        If False, the function returns True/False to be used by the validate address function.
+    """
+    if address_str in NULL_VALUES:
+        return (None, "null") if clean else False
+
+    address_str = address_str.replace("(", "")
```

Comment:
What do you think about `address_str = re.sub(r"[().]", "", address_str)` here? Maybe it's a bit more concise.